Add ARM64 support

glevand opened the pull request:
This series adds arm64 support to rkt.  It also includes some minor build enhancements.  I can split the series into multiple PRs if it would make review easier.

I tested cross building stage1 flavors 'coreos' and 'fly'.  The rkt binary and ACI images generated can run successfully on CoreOS running on qemu-system-aarch64.

On my system (ubuntu 15.10), cross building stage1 flavor 'kvm' failed: ```configure: error: ** No development headers for openssl found```.  I'll look into getting and environment with the needed cross build dependencies.

This PR includes the gimme script submitted upstream at https://github.com/travis-ci/gimme/pull/45. If that is merged we can remove the gimme.local here.

cc: @crawford

glevand commented:
Pushed out updates to fix some test errors.

alban commented:
Thanks for the work on arm64!

This has a lot of Godeps changes but `Godeps/Godeps.json` is not modified by this PR. How did you update it? We use this documentation for updating dependencies:
https://github.com/coreos/rkt/blob/master/Documentation/hacking.md#managing-dependencies
Although this might change with @s-urbaniak's work on Glide (https://github.com/coreos/rkt/pull/2735)

Since it is a lot of changes and GitHub is not able to display the diff, maybe you could write a PR for just Godeps first and get the rest of the changes in a second PR?

glevand commented:
@alban  For the Godeps, I just replaced the files from upstream.  I'll split that into a separate PR and follow the prescribed procedure.

Also, How can I find out why Jenkins is failing?  I seem to need CoreOS login credentials.

glevand commented:
Rebased to latest, removed patch 'Godeps: Update sys/unix to get arm64 support'.

glevand commented:
Depends on https://github.com/coreos/rkt/pull/2764 ```Godeps: Update sys/unix to get arm64 support```.

s-urbaniak commented:
@glevand looks great, can't wait to try it out.

Regarding the jenkins failure, this is the relevant failure, but I don't see how it is related to your PR:
```
17:08:15 === RUN   TestResumedFetch
17:08:15 2016/06/08 17:08:15 http: panic serving 127.0.0.1:46179: send on closed channel
17:08:15 goroutine 211 [running]:
17:08:15 net/http.(*conn).serve.func1(0xc820087880)
17:08:15 	/usr/lib/go/src/net/http/server.go:1389 +0xc1
17:08:15 panic(0x9f8820, 0xc820128950)
17:08:15 	/usr/lib/go/src/runtime/panic.go:426 +0x4e9
17:08:15 github.com/coreos/rkt/tests.testInterruptingServerHandler.func1(0x7f64610c8258, 0xc82033dba0, 0xc82009c540)
17:08:15 	/home/admin/workspace/rkt-matrix/dummy/dummy1/label/debian-8-flavor-coreos/builds/build-rkt-coreos/build-rkt-1.7.0+git/gopath/src/github.com/coreos/rkt/tests/rkt_fetch_test.go:388 +0xced
17:08:15 github.com/coreos/rkt/tests.testServerHandler.func1(0x7f64610c8258, 0xc82033dba0, 0xc82009c540)
17:08:15 	/home/admin/workspace/rkt-matrix/dummy/dummy1/label/debian-8-flavor-coreos/builds/build-rkt-coreos/build-rkt-1.7.0+git/gopath/src/github.com/coreos/rkt/tests/rkt_fetch_test.go:313 +0x63
17:08:15 net/http.HandlerFunc.ServeHTTP(0xc82035c0a0, 0x7f64610c8258, 0xc82033dba0, 0xc82009c540)
17:08:15 	/usr/lib/go/src/net/http/server.go:1618 +0x3a
17:08:15 net/http.serverHandler.ServeHTTP(0xc8200b8100, 0x7f64610c8258, 0xc82033dba0, 0xc82009c540)
17:08:15 	/usr/lib/go/src/net/http/server.go:2081 +0x19e
17:08:15 net/http.(*conn).serve(0xc820087880)
17:08:15 	/usr/lib/go/src/net/http/server.go:1472 +0xf2e
17:08:15 created by net/http.(*Server).Serve
17:08:15 	/usr/lib/go/src/net/http/server.go:2137 +0x44e
17:08:15 --- FAIL: TestResumedFetch (0.34s)
```

iaguis commented:
It's not related, Jenkins is not reliable yet, working on it in https://github.com/coreos/rkt/pull/2762

s-urbaniak commented:
Regarding the Godeps bumps/changes: I went through all commits I think can catch up with this in case it lands first and factor it in my glide PR.

glevand commented:
Rebased to latest, which includes #2764 ```Godeps: Update sys/unix to get arm64 support```.

glevand commented:
Rebased to latest.

lucab commented on stage1/usr_from_coreos/coreos-common.mk:
```diff
@@ -11,9 +11,13 @@ $(call setup-tmp-dir,CCN_TMPDIR)
+CCN_IMG_RELEASE := 1053.2.0
+else
 CCN_IMG_RELEASE := 1032.0.0
+endif
```
General concern: I'm not sure we want to venture into this path of having different arch-specific versions.

glevand replied:
@lucab So you prefer a patch that bumps x86 to 1053.2.0 or later?

krnowak replied:
I'd opt-in for a single version too. But that would need testing if nothing breaks in amd64 too.

lucab replied:
Let's update to 1068.0.0 then, which is also the current base for the beta channel. I'm testing a bump to it right now for amd64, if everything is fine I'll ping back here with the PR.

lucab replied:
PR up at https://github.com/coreos/rkt/pull/2821. @glevand: it shouldn't break arm64 related stuff, but please double-check.

glevand replied:
@lucab 1068.0.0 builds OK on arm64.  I'll try running it and adjust my stage1 manifest files if needed.

glevand replied:
@lucab Tested 1068.0.0 based arm64 stage1-coreos.aci and stage1-fly.aci and they seem to run OK.

tmrts commented:
@coreos/rkt-maintainers PTAL for 1.9.0

krnowak commented on configure.ac:
```diff
@@ -724,7 +735,8 @@ RKT_IF_HAS_FLAVOR([${RKT_STAGE1_FLAVORS}], [coreos,kvm],
         local CoreOS PXE image path:            '${RKT_LOCAL_COREOS_PXE_IMAGE_PATH}'
-        local CoreOS PXE image systemd version: '${RKT_LOCAL_COREOS_PXE_IMAGE_SYSTEMD_VER}'])])
+        local CoreOS PXE image systemd version: '${RKT_LOCAL_COREOS_PXE_IMAGE_SYSTEMD_VER}'
+        stage1 CoreOS board:                    '${RKT_STAGE1_COREOS_BOARD}'])])
```
Just asking: why it is named "CoreOS board"?

glevand replied:
@krnowak 'board' is the terminology used by coreos for the build target.  Currently we have amd64-usr and arm64-usr.

marineam replied:
Specifically inherited from ChromeOS where each image type literally is targeting a specific piece of physical hardware. Our image types are more general purpose so it makes less sense but the term is used so extensively it wouldn't really be worth changing.

krnowak replied:
Alright, thanks for info!

krnowak commented on makelib/variables.mk:
```diff
@@ -24,7 +24,12 @@ QUICKRMTOOL := $(TOOLSDIR)/quickrm
 GO_TEST_FUNC_ARGS ?=
 
-GO_ENV := $(strip GOPATH="$(GOPATH)" $(if $(strip $(GOROOT)),GOROOT=$(strip $(GOROOT))))
+GO_ENV := \
```
Maybe wrap `$(GOARCH)`, `$(CC)` and `$(CXX)` in double quotes and wrap whole assignment in the `$(strip)` function, like in the following:
```
GO_ENV := $(strip \
	GOARCH="$(GOARCH)" \
	CGO_ENABLED=1 \
	CC="$(CC)" \
	CXX="$(CXX)" \
	GOPATH="$(GOPATH)" \
	$(if $(strip $(GOROOT)),GOROOT=$(strip $(GOROOT))))
```
(I hope I balanced the parens correctly here.)

glevand replied:
@krnowak: Should we add quotes around GOROOT also: ```GOROOT="$(strip $(GOROOT))"```?

krnowak replied:
Yes, please.

krnowak commented on stage1/stage1.mk:
```diff
@@ -136,7 +136,7 @@ endif
 $$(call forward-vars,$$(STAGE1_ACI_IMAGE_$1), \
 	ACTOOL STAGE1_ACIDIR_$1)
-$$(STAGE1_ACI_IMAGE_$1): $$(ACTOOL_STAMP) | $$(BINDIR)
+$$(STAGE1_ACI_IMAGE_$1): $$(ACTOOL_STAMP) | $$(TOOLSDIR)
```
Uh, this change is wrong, I think. We have an order-only dependency on `$(BINDIR)` here, because this is where we put our generated stage1 images. So we want to be sure that `$(BINDIR)` exists before we run `actool`.

krnowak commented on makelib/variables.mk:
```diff
@@ -14,7 +14,7 @@ FILELISTDIR := $(BUILDDIR)/filelists
 CLEANDIR := $(BUILDDIR)/clean
 
-ACTOOL := $(BINDIR)/actool
+ACTOOL := $(TOOLSDIR)/actool
```
We have to make sure that release guide is updated about the new location of actool. I don't remember if we run some script to assemble the final tarball with rkt binary and whatnot, but if so, it also needs to be updated.

Other question is whether actool should be distributed at all.

krnowak commented on makelib/variables.mk:
```diff
@@ -6,7 +6,9 @@ override BUILDDIR := $(abspath $(BUILDDIR))
 STAMPSDIR := $(BUILDDIR)/stamps
 TOOLSDIR := $(BUILDDIR)/tools
-BINDIR := $(BUILDDIR)/bin
+TARGETDIR := $(BUILDDIR)/target
```
I was wondering if the target directory can contain an arch in its name, something like `$(BUILDDIR)/target-$(RKT_ACI_ARCH)` or `$(BUILDDIR)/target-$(GOARCH)`.

krnowak replied:
Also, release guide needs to be updated for this too.

krnowak commented on makelib/variables.mk:
```diff
@@ -16,12 +16,16 @@ FILELISTDIR := $(BUILDDIR)/filelists
 QUICKRMTOOL := $(TOOLSDIR)/quickrm
 
+# target tools
+ACTOOL_TARGET := $(TARGET_BINDIR)/actool
```
Ah, this change can be actually dropped. We don't distribute actool in tarballs. If somebody wants to have an `actool` binary, then they should just build it from appc/spec.

krnowak replied:
I meant - this commit can be dropped.

glevand replied:
OK.  CoreOS currently builds and installs actool for the target along with rkt, and is why I added this.  I'll look into adding a new CoreOS ebuild for actool.

krnowak commented on stage1/stage1.mk:
```diff
@@ -66,7 +66,7 @@ $(foreach f,$(STAGE1_FLAVORS), \
 	$(eval STAGE1_ACIDIR_$f := $(BUILDDIR)/aci-for-$f-flavor) \
 	$(eval STAGE1_ACIROOTFSDIR_$f := $(STAGE1_ACIDIR_$f)/rootfs) \
-	$(eval STAGE1_ACI_IMAGE_$f := $(BINDIR)/stage1-$f.aci) \
+	$(eval STAGE1_ACI_IMAGE_$f := $(TARGET_BINDIR)/stage1-$f.aci) \
```
In the same file, there is a rule that begins with `$$(STAGE1_ACI_IMAGE_$1): $$(ACTOOL_STAMP) |`. You should update the directory there too to `$(TARGET_BINDIR)`.

krnowak commented on stage1_fly/aci/aci.mk:
```diff
@@ -25,18 +26,19 @@ FLY_ACI_DIRS := \
-# invalidate generated manifest if version changes
-$(call generate-kv-deps,$(FLY_ACI_MANIFEST_KV_DEPMK_STAMP),$(FLY_ACI_GEN_MANIFEST),$(FLY_ACI_MANIFEST_KV_DEPMK),FLY_ACI_VERSION)
+# invalidate generated manifest if version or arch changes
+$(call generate-kv-deps,$(FLY_ACI_MANIFEST_KV_DEPMK_STAMP),$(FLY_ACI_GEN_MANIFEST),$(FLY_ACI_MANIFEST_KV_DEPMK),FLY_ACI_VERSION,FLY_ACI_ARCH)
```
`FLY_ACI_VERSION` and `FLY_ACI_ARCH` should be separated with a space, not comma.

krnowak commented on .travis.yml:
```diff
@@ -17,13 +26,22 @@ addons:
+                eval "$(GIMME_ARCH=${TARGET} GIMME_CGO_ENABLED=1 ./gimme.local ${TRAVIS_GO_VERSION})";
+                ./autogen.sh;
+                ./configure --host=aarch64-linux-gnu --with-stage1-flavors=host,fly --enable-tpm=auto;
+                make V=2;
```
Shouldn't this be `make check` too?

glevand replied:
I get this when I try: ```go test: -race is only supported on linux/amd64, freebsd/amd64, darwin/amd64 and windows/amd64```.

I plan to setup our local Jenkins to do tests on an arm64 platform.  Do you have any other suggestions?

krnowak replied:
I'm not sure what travis is doing here, what does it do when you tell it to build arm64 software - does it do the cross-compilation or does it the build on arm64 hardware/VM?

So, I'm not sure, maybe replace `--race` in `tests/tests.mk` with `$(if $(filter $(GOARCH),amd64),--race)`? So it will pass --race only when GOARCH is amd64.

OTOH, it might not be what we want, right?

lucab replied:
Looking at the rest of this PR, it looks like it is just cross-compiling. I don't think we will be able to run unit-tests there, right?

glevand replied:
Yes, it is cross compiled.  Static code analysis tools may work, but we cannot do run-time testing.

krnowak commented:
There are some changes in the builddir layout that probably require some change log entry too, as the built rkt and stage1 images are now in a different directory.

krnowak commented:
CCing myself, @krnowak

glevand commented:
Rebased to latest, addressed all comments so far.

krnowak commented on CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+
+#### Other changes
+
+- build directory layout changes (): The rkt and stage1 image files have been moved from the 'bin' sub-directory to the 'target/bin' sub-directory.
```
What are these empty parens for?

glevand replied:
I'll put the pull request link in there.

krnowak replied:
Well, you already know the link. :)

krnowak commented on stage1/find-em.sh:
```diff
@@ -0,0 +1,134 @@
+
+: ${BOARD:="arm64-usr"}
+if [ "${BOARD}" = "arm64-usr" ] ; then
+	: ${VERSION:="1053.2.0"}
```
1068.0.0

krnowak commented on .travis.yml:
```diff
@@ -8,16 +8,40 @@ go:
+
+matrix:
+  exclude:
+    - go: 1.6
```
Is something wrong with this combination?

glevand replied:
No, I thought to put it in to speed up the travis job.  I'll remove this exclude.

krnowak commented on stage1/init/init_linux_amd64.go:
```diff
@@ -0,0 +1,22 @@
+// Copyright 2014 The rkt Authors
```
2016.

krnowak commented on stage1/init/init_linux_arm64.go:
```diff
@@ -0,0 +1,22 @@
+// Copyright 2014 The rkt Authors
```
2016

krnowak commented on stage1/find-em.sh:
```diff
@@ -0,0 +1,134 @@
+#!/bin/sh
```
Consider moving this script to scripts directory.

krnowak commented on stage1/find-em.sh:
```diff
@@ -0,0 +1,134 @@
+#
+
+# TODO: get values from makefile
+#: ${BOARD:="@@BOARD@@"}
```
Untested.

```bash
mkvar () {
	local mkfile="$1"
	local varname="$2"

	grep -e "${varname}\s*:=\s*" "${mkfile}" | cut -d= -f2- | sed -e 's/^\s\+//'
}

d=$(dirname $0)
BOARD=$(mkvar "$d/../Makefile" 'RKT_STAGE1_COREOS_BOARD')
VERSION=$(mkvar "$d/../stage1/usr_from_coreos/coreos-common.mk" 'CCN_IMG_RELEASE')
```

glevand replied:
With that we'd need the makefiles on the target.  I was thinking to use sed in the makefile to generate the script, then it will be self contained.

krnowak replied:
Well, if makefile will generate the script, then you would still have the generated makefile, wouldn't you? :) Either way, you have to call `./configure`.

What you could do is to add a command line parameter that will have a precendence over the value in the makefile, so it won't be strictly necessary to have a generated make file first.

krnowak commented:
Some more minor nits.

Also, the release guide needs updating - the file is in `Documentation/devel/release.md`. Basically all instances of `release-build/bin` or `release-build/tools` need to be checked and probably changed to `release-build/target/bin` and `release-build/target/tools`, respectively.

Also, when this is merged, we probably should release tarballs with arm64 binaries too, but this is something to discuss in a separate issue.

lucab commented:
> Also, the release guide needs updating - the file is in Documentation/devel/release.md. Basically all instances of release-build/bin or release-build/tools need to be checked and probably changed to release-build/target/bin and release-build/target/tools, respectively.

As this keeps getting bigger, do we want to split the multiarch-ification first out of this and make sure that everything keeps working under amd64? There are also some minor build-infra cleanup which can get in as well, leaving here only the core of arm64 stuff.

> Also, when this is merged, we probably should release tarballs with arm64 binaries too, but this is something to discuss in a separate issue.

I guess we may want to have release autobuilders at that point, avoiding local builds of release artifacts.

krnowak commented:
>> Also, the release guide needs updating - the file is in Documentation/devel/release.md. Basically all instances of release-build/bin or release-build/tools need to be checked and probably changed to release-build/target/bin and release-build/target/tools, respectively.
> 
> As this keeps getting bigger, do we want to split the multiarch-ification first out of this and make sure that everything keeps working under amd64? There are also some minor build-infra cleanup which can get in as well, leaving here only the core of arm64 stuff.

I only asked here for some renames as built artifacts are now in a different place.

glevand commented:
Rebased to latest.

I added a new top level makefile variable RKT_STAGE1_COREOS_VERSION to go along with RKT_STAGE1_COREOS_BOARD.  The two together ```$(RKT_STAGE1_COREOS_BOARD)/$(RKT_STAGE1_COREOS_VERSION)``` define the CoreOS release used.

I moved find-em.sh to the tools directory and added a makefile for it.  The tools directory seemed like the place for it.  I can move it to the scripts directory and add a scripts.mk if you think it better.  The unprocessed find-em input file can still be used with something like ```RELEASE="arm64-usr/1068.0.0" find-em.sh.in```.

Updated 	docs release.md and update-coreos-stage1.md.

krnowak commented on Makefile.in:
```diff
@@ -44,6 +44,7 @@ RKT_LOCAL_COREOS_PXE_IMAGE_PATH := @RKT_LOCAL_COREOS_PXE_IMAGE_PATH@
 RKT_LOCAL_COREOS_PXE_IMAGE_SYSTEMD_VER := @RKT_LOCAL_COREOS_PXE_IMAGE_SYSTEMD_VER@
 
 RKT_STAGE1_COREOS_BOARD := @RKT_STAGE1_COREOS_BOARD@
+RKT_STAGE1_COREOS_VERSION := @RKT_STAGE1_COREOS_VERSION@
```
I'm not sure I'm happy with this change:

1. It isn't really a general value, its use is contained entirely inside coreos-common.mk.
2. Now, if we want to upgrade the version of coreos image, we have to do the modifications in both configure.ac and coreos-common.mk.
3. Might be misleading, because it will be ignored if we use the `--with-coreos-local-pxe-image-*` flags.

krnowak commented on tools/find-em.sh.in:
```diff
@@ -0,0 +1,126 @@
+
+	if [ ! -f ${file} ]; then
+		echo "${url}:"
+		curl --progress-bar -o ${file} ${url}
```
I'd prefer `wget` - we already use it elsewhere.

krnowak replied:
Anyway, I'd like to propose to move this script to a separate PR. I'd like this PR to merged ASAP, but the likely bikeshedding over this script will only delay it.

krnowak commented on Documentation/devel/update-coreos-stage1.md:
```diff
@@ -1,26 +1,34 @@
+## Generate the manifest entries
 
-Let's assume you want to update from version 991.0.0 to version 1032.0.0.
+Two methods to generate the entries of the manifest files are outlined here.  The find-em script is run from the CoreOS relese of interest and generates candidate entries.  The image extraction method is static and does not require a running CoreOS relese.
```
s/relese/release

But I'd like to move the find-em thing to a separate PR.

krnowak commented on Documentation/devel/update-coreos-stage1.md:
```diff
@@ -64,7 +72,7 @@ You should have now the rootfs of the image in the `squashfs-root` directory.
 ## Update the manifest files
 
-Back to the rkt repo, in the directory `stage1/usr_from_coreos/manifest.d`, there are some manifest files that define which files are copied from the CoreOS image to the stage1 image.
+Back to the rkt repo, in the directory `stage1/usr_from_coreos/manifest-XXX.d`, there are some manifest files that define which files are copied from the CoreOS image to the stage1 image.
```
Explain please what this `XXX` is.

krnowak commented on Documentation/devel/update-coreos-stage1.md:
```diff
@@ -73,31 +81,40 @@ In our case, there are no updates and all the files mentioned in the manifest ar
 
-In the file `stage1/usr_from_coreos/coreos-common.mk`, we define which CoreOS image version we use for the coreos flavor.
-Update `CCN_IMG_RELEASE` to 1032.0.0 and `CCN_SYSTEMD_VERSION` to the systemd version shipped with the image (in our case, v229).
+In the files `configure.ac` and `stage1/usr_from_coreos/coreos-common.mk`, we define which CoreOS image version and systemd version we use for the coreos flavor.
```
As I said, I'm not sure I like this change.

glevand commented:
Rebased to latest (v1.9.0).

Removed the find-em patch and the related RKT_STAGE1_COREOS_VERSION and update-coreos-stage1.md patches.

glevand commented:
Rebased to latest.

krnowak commented:
LFAD.

jonboulle commented:
@glevand thanks for the awesome work and your patience!
